pkg/utils/proc/ns.go

```diff
@@ -0,0 +1,163 @@
+package proc
+
+import (
+	"bytes"
+	"fmt"
+	"os"
+	"strconv"
+	"strings"
+
+	"github.com/aquasecurity/tracee/pkg/logger"
+)
+
+// GetMountNSFirstProcesses return mapping between mount NS to its first process
+// (aka, the process with the oldest start time in the mount NS)
+func GetMountNSFirstProcesses() (map[int]int, error) {
+	procDir, err := os.Open("/proc")
+	if err != nil {
+		return nil, fmt.Errorf("could not open proc dir: %v", err)
+	}
+	defer procDir.Close()
+
+	entries, err := procDir.Readdirnames(-1)
+	if err != nil {
+		return nil, fmt.Errorf("could not read proc dir: %v", err)
+	}
+
+	type pidTimestamp struct {
+		pid       uint
+		timestamp int
+	}
+	mountNSTimeMap := make(map[int]pidTimestamp)
+	// Iterate over each pid
+	for _, entry := range entries {
+		pid, err := strconv.ParseUint(entry, 10, 32)
+		if err != nil {
+			continue
+		}
+		procNS, err := GetProcNS(uint(pid))
+		if err != nil {
+			logger.Warn("Failed in fetching process mount namespace", pid, err.Error())
+			continue
+		}
+
+		processStartTime, err := GetProcessStartTime(uint(pid))
+		if err != nil {
+			logger.Warn("Failed in fetching process start time", pid, err.Error())
+			continue
+		}
+
+		currentNSProcess, ok := mountNSTimeMap[procNS.Mnt]
+		if ok {
+			// If executed after current save process, it can't be the first process
+			if processStartTime >= currentNSProcess.timestamp {
+				continue
+			}
+		}
+		mountNSTimeMap[procNS.Mnt] = pidTimestamp{timestamp: processStartTime, pid: uint(pid)}
+	}
+
+	mountNSToFirstProcess := make(map[int]int)
+	for mountNS, p := range mountNSTimeMap {
+		mountNSToFirstProcess[mountNS] = int(p.pid)
+	}
+	return mountNSToFirstProcess, nil
+}
+
+// GetProcessStartTime return the start time of the process using the procfs
+func GetProcessStartTime(pid uint) (int, error) {
+	stat, err := os.ReadFile(fmt.Sprintf("/proc/%d/stat", pid))
+	if err != nil {
+		return 0, err
+	}
+	// see https://man7.org/linux/man-pages/man5/proc.5.html for how to read /proc/pid/stat
+	startTimeOffset := 22 // Offset start at 1
+	splitStat := bytes.SplitN(stat, []byte{' '}, startTimeOffset+1)
+	if len(splitStat) != startTimeOffset+1 {
+		return 0, fmt.Errorf("error in parsing /proc/<pid>/stat format - only %d values found inside", len(splitStat))
+	}
+	startTime, err := strconv.Atoi(string(splitStat[startTimeOffset-1]))
+	if err != nil {
+		return 0, err
+	}
+
+	return startTime, nil
+}
+
+type ProcNS struct {
+	Cgroup          int
+	Ipc             int
+	Mnt             int
+	Net             int
+	Pid             int
+	PidForChildren  int
+	Time            int
+	TimeForChildren int
+	User            int
+	Uts             int
+}
+
+// GetProcNS return all the namespaces of a given process.
+// To do so, it requires access to the /proc file system of the host, and CAP_SYS_PTRACE capability.
```

Comment:
If CAP_SYS_PTRACE is required, shouldn't we request the relevant privilege here?
Also, shouldn't we add it to the dependencies?

Reply:
Yes, we do:

```
{"level":"warn","ts":1669382671.3352199,"msg":"Failed in fetching process mount namespace"}
{"level":"error","ts":1669382671.3352795,"msg":"Ignored key-value pairs with non-string keys.","invalid":[{"position":0,"key":1926613,"value":"readlink /proc/1926613/ns/net: permission denied"}]}
{"level":"warn","ts":1669382671.3352783,"msg":"Failed in fetching process mount namespace"}
{"level":"error","ts":1669382671.3353367,"msg":"Ignored key-value pairs with non-string keys.","invalid":[{"position":0,"key":1926614,"value":"readlink /proc/1926614/ns/net: permission denied"}]}
{"level":"warn","ts":1669382671.3353357,"msg":"Failed in fetching process mount namespace"}
```

when running with `--capabilities bypass=false`

Reply:
But it is not only CAP_SYS_PTRACE, since with:

```diff
--- a/pkg/ebpf/tracee.go
+++ b/pkg/ebpf/tracee.go
@@ -468,7 +468,12 @@ func (t *Tracee) Init() error {
                t.config.maxPidsCache = 5 // default value for config.maxPidsCache
        }
        t.pidsInMntns.Init(t.config.maxPidsCache)
-       mntNSProcs, err := proc.GetMountNSFirstProcesses()
+
+       var mntNSProcs map[int]int
+       err = capabilities.GetInstance().Requested(func() error {
+               mntNSProcs, err = proc.GetMountNSFirstProcesses()
+               return err
+       }, cap.SYS_PTRACE)
        if err == nil {
                for mountNS, pid := range mntNSProcs {
                        t.pidsInMntns.AddBucketItem(uint32(mountNS), uint32(pid))
```

There are still errors but when I replace with:

```diff
-       mntNSProcs, err := proc.GetMountNSFirstProcesses()
+
+       var mntNSProcs map[int]int
+       err = capabilities.GetInstance().Privileged(func() error {
+               mntNSProcs, err = proc.GetMountNSFirstProcesses()
+               return err
+       })
```

it works. Need to find correct capabilities for the block.

Reply:
I also discovered that the procInfo initialization fails with:

```
could not parse task status: readlink /proc/1822587/task/1822587/ns/mnt: permission deniedpid: 1822590
could not parse task status: readlink /proc/1822590/task/1822590/ns/mnt: permission deniedpid: 1822591
could not parse task status: readlink /proc/1822591/task/1822591/ns/mnt: permission deniedpid: 1822635
```

but it was failing silently when dropping capabilities because the errors are always continued (as we didn't have a logger/debug back then).

Reply:
Alright, I found all needed capabilities, proposing a PR.

Reply:
Fixes at: https://github.com/aquasecurity/tracee/pull/2380